meadow-client/src/Semantics.purs

```diff
@@ -1159,88 +1154,67 @@ applyTransaction inputs sigs blockNum state contract value = case appResult of
   appResult = applyAnyInputs inputs sigs neededInputs blockNum expiredState reducedContract value emptyOutcome Nil
 
 -- Extract participants from state and contract
-peopleFromCommitInfo ::
-  CommitInfo ->
-  S.Set Person
-peopleFromCommitInfo (CommitInfo { redeemedPerPerson: rpp, currentCommitsById: ccbi }) = S.union (S.fromFoldable (M.keys rpp)) (S.fromFoldable (map (\x ->
-  x.person) (M.values ccbi)))
-
-peopleFromState :: State -> S.Set Person
-peopleFromState (State { commits: comm }) = peopleFromCommitInfo comm
-
-peopleFromValue :: Value -> S.Set Person
-peopleFromValue CurrentBlock = S.empty
-
-peopleFromValue (Committed _) = S.empty
-
-peopleFromValue (Constant _) = S.empty
-
-peopleFromValue (NegValue value) = peopleFromValue value
-
-peopleFromValue (AddValue value1 value2) = S.union (peopleFromValue value1) (peopleFromValue value2)
-
-peopleFromValue (SubValue value1 value2) = S.union (peopleFromValue value1) (peopleFromValue value2)
-
-peopleFromValue (MulValue value1 value2) = S.union (peopleFromValue value1) (peopleFromValue value2)
-
-peopleFromValue (DivValue value1 value2 value3) = S.union (peopleFromValue value1) (S.union (peopleFromValue value2) (peopleFromValue value3))
-
-peopleFromValue (ModValue value1 value2 value3) = S.union (peopleFromValue value1) (S.union (peopleFromValue value2) (peopleFromValue value3))
-
-peopleFromValue (ValueFromChoice v value) = S.insert (unwrap v).person (peopleFromValue value)
-
-peopleFromValue (ValueFromOracle _ value) = peopleFromValue value
-
-peopleFromObservation :: Observation -> S.Set Person
-peopleFromObservation (BelowTimeout _) = S.empty
-
-peopleFromObservation (AndObs observation1 observation2) = S.union (peopleFromObservation observation1) (peopleFromObservation observation2)
-
-peopleFromObservation (OrObs observation1 observation2) = S.union (peopleFromObservation observation1) (peopleFromObservation observation2)
-
-peopleFromObservation (NotObs observation) = peopleFromObservation observation
-
-peopleFromObservation (ChoseThis v _) = S.insert (unwrap v).person S.empty
+class PeopleFrom a where
+  peopleFrom :: a -> S.Set Person
 
-peopleFromObservation (ChoseSomething v) = S.insert (unwrap v).person S.empty
+instance peopleFromCommitInfo :: PeopleFrom CommitInfo where
+  peopleFrom (CommitInfo { redeemedPerPerson: rpp, currentCommitsById: ccbi })
+    = S.union (S.fromFoldable (M.keys rpp)) (S.fromFoldable (map (\x -> x.person) (M.values ccbi)))
```

Comment:
`foldMap _.person (M.values ccbi)`?

Reply:
not quite, I think you still need to turn it into a `Set` somehow.

Reply:
Ah yes. `foldMap (_.person >>> singleton) (M.values ccbi)`? 🤔